File: ops/ascendc/meta_gauss_render/csrc/CalcRender.cpp
/**
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */
#include <string>
#include <torch/torch.h>
#include "functions.h"
#include "OpApiCommon.h"

using namespace NPU_NAME_SPACE;
using namespace std;

at::Tensor calc_render_bwd_var_clip_gsids(
    const at::Tensor &vColor, const at::Tensor &vDepth, const at::Tensor &lastCumsum, const at::Tensor &error,
    const at::Tensor &gs, const at::Tensor &tileCoords,
    const at::Tensor &offsets, const at::Tensor &gsIds, const at::Tensor &gsClipIndex, const at::Tensor &alphaClipIndex)
{
    TORCH_CHECK(vColor.device().type() == at::kPrivateUse1, "Invalid device.");
    TORCH_CHECK(vDepth.device() == vColor.device(), "Inconsistent device.");
    TORCH_CHECK(lastCumsum.device() == vColor.device(), "Inconsistent device.");
    TORCH_CHECK(error.device() == vColor.device(), "Inconsistent device.");

    TORCH_CHECK(gs.device() == vColor.device(), "Inconsistent device.");
    TORCH_CHECK(tileCoords.device() == vColor.device(), "Inconsistent device.");
    TORCH_CHECK(offsets.device() == vColor.device(), "Inconsistent device.");
    TORCH_CHECK(gsIds.device() == vColor.device(), "Inconsistent device.");
    TORCH_CHECK(gsClipIndex.device() == vColor.device(), "Inconsistent device.");
    TORCH_CHECK(alphaClipIndex.device() == vColor.device(), "Inconsistent device.");

    auto device = vColor.device();
    auto options = at::TensorOptions().dtype(at::kFloat).layout(at::kStrided).device(device);

    int64_t nGauss = gs.sizes()[1];
    int64_t nPixel = vColor.sizes()[2];

    at::Tensor vGs = at::zeros_like(gs, options);
    at::Tensor gsClipIndex_gsIds = at::cat({gsClipIndex, gsIds}, 0); // dim = 0

    EXEC_NPU_CMD(aclnnCalcRenderBwdVarClipGsids,
        vColor, vDepth, lastCumsum, error,
        gs, tileCoords, offsets, gsClipIndex_gsIds,
        alphaClipIndex, vGs);
    return vGs;
}

std::tuple<at::Tensor, at::Tensor, at::Tensor, at::Tensor, at::Tensor, at::Tensor> calc_render_fwd_double_clip_gsids(
    const at::Tensor &gs, const at::Tensor &tileCoords, const at::Tensor &offsets, const at::Tensor &gsIds)
{
    TORCH_CHECK(gs.device().type() == at::kPrivateUse1, "Invalid device.");
    TORCH_CHECK(tileCoords.device() == gs.device(), "Inconsistent device.");
    TORCH_CHECK(offsets.device() == gs.device(), "Inconsistent device.");
    TORCH_CHECK(gsIds.device() == gs.device(), "Inconsistent device.");
    
    auto device = gs.device();
    auto options = at::TensorOptions().dtype(at::kFloat).layout(at::kStrided).device(device);

    int64_t tileNum = tileCoords.sizes()[0];
    int64_t nPixel = tileCoords.sizes()[2];

    at::Tensor color = at::zeros({3, tileNum, nPixel}, options);
    at::Tensor depth = at::zeros({1, tileNum, nPixel}, options);
    at::Tensor lastCumsum = at::zeros({tileNum, nPixel}, options);
    at::Tensor error = at::zeros({tileNum, nPixel}, options);
    
    options = at::TensorOptions().dtype(torch::kInt64).layout(at::kStrided).device(device);
    at::Tensor gsClipIndex = at::zeros({tileNum}, options);
    options = at::TensorOptions().dtype(torch::kUInt8).layout(at::kStrided).device(device);
    at::Tensor alphaClipIndex = at::zeros({gsIds.sizes()[0], 2}, options);

    EXEC_NPU_CMD(aclnnCalcRenderFwdDoubleClipGsids,
        gs, tileCoords, offsets, gsIds,
        color, depth, lastCumsum, error,
        gsClipIndex, alphaClipIndex);
    return std::tie(color, depth, lastCumsum, error, gsClipIndex, alphaClipIndex);
}

File: ops/ascendc/meta_gauss_render/ops/calc_render.py
# coding=utf-8
# This program is free software, you can redistribute it and/or modify it under the terms and conditions of
# CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.

import torch
import meta_gauss_render._C


# pylint: disable=too-many-arguments,huawei-too-many-arguments,too-many-return-values
class CalcRender(torch.autograd.Function):
    """transmittance"""

    @staticmethod
    def forward(
        ctx, means, conic0s, conic1s, conic2s, opacities, colors, depths, tile_coords, offsets, sorted_gs_ids
    ):
        depths_ = depths
        if depths is None:
            depths_ = colors[0:1]
        sorted_gs_ids = sorted_gs_ids.to(torch.int64)
        # tile size < 64 clip
        gs = torch.cat([
            means,
            conic0s.unsqueeze(0),
            conic1s.unsqueeze(0),
            conic2s.unsqueeze(0),
            opacities.unsqueeze(0),
            colors,
            depths_
            ], dim=0).T.contiguous()
        color, depth, last_cumsum, error, gs_clip_index, alpha_clip_index = \
            meta_gauss_render._C.calc_render_fwd_double_clip_gsids(gs, tile_coords, offsets, sorted_gs_ids)
        
        ctx.save_for_backward(gs, depths, tile_coords, offsets, sorted_gs_ids,
            last_cumsum, error, gs_clip_index, alpha_clip_index)

        if depths is None:
            return color
        else:
            return color, depth

    @staticmethod
    def backward(
        ctx, *v_args
    ):
        v_color, v_depth = None, None
        if len(v_args) == 2:
            v_color, v_depth = v_args
        elif len(v_args) == 1:
            v_color = v_args
            v_depth = torch.zeros_like(v_color[0:1])
        else:
            raise Exception("invalid arguments")
        v_color = v_color.contiguous()
        v_depth = v_depth.contiguous()

        gs, depths, tile_coords, offsets, sorted_gs_ids, last_cumsum, error, gs_clip_index, \
            alpha_clip_index = ctx.saved_tensors

        v_gs = meta_gauss_render._C.calc_render_bwd_var_clip_gsids(
            v_color, v_depth, last_cumsum, error,
            gs,
            tile_coords, offsets,
            sorted_gs_ids,
            gs_clip_index,
            alpha_clip_index
        )
        v_means = v_gs[:, :2].T         # shape: (2, N)
        v_conic0s = v_gs[:, 2:3].T        # shape: (1, N)
        v_conic1s = v_gs[:, 3:4].T        # shape: (1, N)
        v_conic2s = v_gs[:, 4:5].T        # shape: (1, N)
        v_opacities = v_gs[:, 5:6].T        # shape: (1, N)
        v_colors = v_gs[:, 6:9].T        # shape: (3, N)

        if depths is None:
            return v_means, v_conic0s, v_conic1s, v_conic2s, v_opacities, v_colors, None, None, None, None
        else:
            v_depths = v_gs[:, 9:].T
            return v_means, v_conic0s, v_conic1s, v_conic2s, v_opacities, v_colors, v_depths, None, None, None


calc_render = CalcRender.apply